File: SS/Labs/Lab8/ll1_lab8.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define MAX_NT 10
#define MAX_PROD 20
#define MAX_RHS 20

typedef struct {
    char lhs;
    char rhs[MAX_RHS];
} Production;

/* Use # for epsilon */
static Production prods[] = {
    {'S', "aBDh"},
    {'B', "cC"},
    {'C', "bC"},
    {'C', "#"},
    {'D', "EF"},
    {'E', "g"},
    {'E', "#"},
    {'F', "f"},
    {'F', "#"}
};

static const int P = sizeof(prods) / sizeof(prods[0]);
static const char nonterms[] = {'S', 'B', 'C', 'D', 'E', 'F'};
static const int N = sizeof(nonterms) / sizeof(nonterms[0]);

static int nt_index(char nt) {
    int i;
    for (i = 0; i < N; i++) {
        if (nonterms[i] == nt) return i;
    }
    return -1;
}

static int add_to_set(int set[128], char c) {
    unsigned char uc = (unsigned char)c;
    if (!set[uc]) {
        set[uc] = 1;
        return 1;
    }
    return 0;
}

static int union_set(int dest[128], const int src[128], int skip_epsilon) {
    int changed = 0;
    int i;
    for (i = 0; i < 128; i++) {
        if (!src[i]) continue;
        if (skip_epsilon && i == '#') continue;
        if (!dest[i]) {
            dest[i] = 1;
            changed = 1;
        }
    }
    return changed;
}

/* FIRST sets for non-terminals */
static int first_nt[MAX_NT][128];
/* FOLLOW sets for non-terminals */
static int follow_nt[MAX_NT][128];

static void first_of_string(const char *str, int out[128]) {
    int i;
    int all_can_be_epsilon = 1;
    memset(out, 0, 128 * sizeof(int));

    if (str[0] == '\0') {
        out['#'] = 1;
        return;
    }

    for (i = 0; str[i] != '\0'; i++) {
        char sym = str[i];

        if (!isupper((unsigned char)sym)) {
            add_to_set(out, sym);
            all_can_be_epsilon = 0;
            break;
        } else {
            int idx = nt_index(sym);
            if (idx < 0) {
                all_can_be_epsilon = 0;
                break;
            }

            union_set(out, first_nt[idx], 1);
            if (!first_nt[idx]['#']) {
                all_can_be_epsilon = 0;
                break;
            }
        }
    }

    if (all_can_be_epsilon) {
        add_to_set(out, '#');
    }
}

static void compute_first_sets(void) {
    int changed = 1;
    while (changed) {
        int i;
        changed = 0;

        for (i = 0; i < P; i++) {
            int lhs_i = nt_index(prods[i].lhs);
            int f[128];

            first_of_string(prods[i].rhs, f);
            if (union_set(first_nt[lhs_i], f, 0)) {
                changed = 1;
            }
        }
    }
}

static void compute_follow_sets(void) {
    int changed = 1;

    /* Start symbol: S */
    add_to_set(follow_nt[nt_index('S')], '$');

    while (changed) {
        int i;
        changed = 0;

        for (i = 0; i < P; i++) {
            int lhs_i = nt_index(prods[i].lhs);
            const char *rhs = prods[i].rhs;
            int j;

            for (j = 0; rhs[j] != '\0'; j++) {
                char B = rhs[j];
                int b_i;

                if (!isupper((unsigned char)B)) continue;
                b_i = nt_index(B);

                if (rhs[j + 1] != '\0') {
                    int first_beta[128];
                    first_of_string(rhs + j + 1, first_beta);

                    if (union_set(follow_nt[b_i], first_beta, 1)) {
                        changed = 1;
                    }
                    if (first_beta['#']) {
                        if (union_set(follow_nt[b_i], follow_nt[lhs_i], 0)) {
                            changed = 1;
                        }
                    }
                } else {
                    if (union_set(follow_nt[b_i], follow_nt[lhs_i], 0)) {
                        changed = 1;
                    }
                }
            }
        }
    }
}

static void print_set(const int set[128]) {
    int first = 1;
    int c;
    printf("{ ");
    for (c = 0; c < 128; c++) {
        if (!set[c]) continue;
        if (!first) printf(", ");
        printf("%c", c);
        first = 0;
    }
    printf(" }");
}

static int terminals[128];
static char term_list[32];
static int T = 0;

static void collect_terminals(void) {
    int i, j;
    memset(terminals, 0, sizeof(terminals));

    for (i = 0; i < P; i++) {
        for (j = 0; prods[i].rhs[j] != '\0'; j++) {
            char c = prods[i].rhs[j];
            if (!isupper((unsigned char)c) && c != '#') {
                terminals[(unsigned char)c] = 1;
            }
        }
    }
    terminals[(unsigned char)'$'] = 1;

    T = 0;
    for (i = 0; i < 128; i++) {
        if (terminals[i]) {
            term_list[T++] = (char)i;
        }
    }
}

static int table_prod[MAX_NT][128];

static void build_ll1_table(int *is_ll1) {
    int i, j;
    *is_ll1 = 1;

    for (i = 0; i < MAX_NT; i++) {
        for (j = 0; j < 128; j++) {
            table_prod[i][j] = -1;
        }
    }

    for (i = 0; i < P; i++) {
        int a_i = nt_index(prods[i].lhs);
        int first_alpha[128];
        first_of_string(prods[i].rhs, first_alpha);

        for (j = 0; j < 128; j++) {
            if (j == '#') continue;
            if (!first_alpha[j]) continue;

            if (table_prod[a_i][j] != -1 && table_prod[a_i][j] != i) {
                *is_ll1 = 0;
            }
            table_prod[a_i][j] = i;
        }

        if (first_alpha['#']) {
            int c;
            for (c = 0; c < 128; c++) {
                if (!follow_nt[a_i][c]) continue;
                if (table_prod[a_i][c] != -1 && table_prod[a_i][c] != i) {
                    *is_ll1 = 0;
                }
                table_prod[a_i][c] = i;
            }
        }
    }
}

static void print_table(void) {
    int i, j;
    printf("\nLL(1) Parsing Table\n\n");
    printf("NT/T\t");
    for (j = 0; j < T; j++) {
        printf("%c\t", term_list[j]);
    }
    printf("\n");

    for (i = 0; i < N; i++) {
        printf("%c\t", nonterms[i]);
        for (j = 0; j < T; j++) {
            int p = table_prod[i][(unsigned char)term_list[j]];
            if (p == -1) {
                printf("-\t");
            } else {
                printf("%c->%s\t", prods[p].lhs, prods[p].rhs);
            }
        }
        printf("\n");
    }
}

int main(void) {
    int i;
    int is_ll1;

    printf("Given Grammar:\n");
    printf("S -> aBDh\n");
    printf("B -> cC\n");
    printf("C -> bC | #\n");
    printf("D -> EF\n");
    printf("E -> g | #\n");
    printf("F -> f | #\n");

    compute_first_sets();
    compute_follow_sets();
    collect_terminals();
    build_ll1_table(&is_ll1);

    printf("\nFIRST Sets:\n");
    for (i = 0; i < N; i++) {
        printf("FIRST(%c) = ", nonterms[i]);
        print_set(first_nt[i]);
        printf("\n");
    }

    printf("\nFOLLOW Sets:\n");
    for (i = 0; i < N; i++) {
        printf("FOLLOW(%c) = ", nonterms[i]);
        print_set(follow_nt[i]);
        printf("\n");
    }

    print_table();

    printf("\nGrammar status: %s\n", is_ll1 ? "Grammar is LL(1)." : "Grammar is NOT LL(1). (Conflict found)");
    return 0;
}

File: SS/Labs/PythonLabs/Lab8/ll1_lab8.py
# Lab 8 — LL(1) Parser with FIRST/FOLLOW computation
# Grammar:
#   S -> aBDh
#   B -> cC
#   C -> bC | #   (# = epsilon)
#   D -> EF
#   E -> g | #
#   F -> f | #

PRODS = [
    ('S', 'aBDh'),
    ('B', 'cC'),
    ('C', 'bC'),
    ('C', '#'),
    ('D', 'EF'),
    ('E', 'g'),
    ('E', '#'),
    ('F', 'f'),
    ('F', '#'),
]

NONTERMS = ['S', 'B', 'C', 'D', 'E', 'F']

def nt_index(c):
    return NONTERMS.index(c) if c in NONTERMS else -1

def is_upper(c):
    return c.isupper()

# ── FIRST of a string ────────────────────────────────────────────────────

def first_of_str(s, first_sets):
    out = set()
    if not s or s == '#':
        return {'#'}
    all_eps = True
    for sym in s:
        if not is_upper(sym):
            out.add(sym)
            all_eps = False
            break
        idx = nt_index(sym)
        if idx == -1:
            all_eps = False
            break
        out |= first_sets[idx] - {'#'}
        if '#' not in first_sets[idx]:
            all_eps = False
            break
    if all_eps:
        out.add('#')
    return out

# ── Compute FIRST sets ───────────────────────────────────────────────────

def compute_first():
    first = [set() for _ in NONTERMS]
    changed = True
    while changed:
        changed = False
        for lhs, rhs in PRODS:
            i = nt_index(lhs)
            f = first_of_str(rhs, first)
            before = len(first[i])
            first[i] |= f
            if len(first[i]) != before:
                changed = True
    return first

# ── Compute FOLLOW sets ──────────────────────────────────────────────────

def compute_follow(first):
    follow = [set() for _ in NONTERMS]
    follow[nt_index('S')].add('$')
    changed = True
    while changed:
        changed = False
        for lhs, rhs in PRODS:
            li = nt_index(lhs)
            for j, sym in enumerate(rhs):
                if not is_upper(sym):
                    continue
                bi = nt_index(sym)
                if bi == -1:
                    continue
                beta = rhs[j+1:]
                fb = first_of_str(beta, first) if beta else {'#'}
                before = len(follow[bi])
                follow[bi] |= fb - {'#'}
                if '#' in fb:
                    follow[bi] |= follow[li]
                if len(follow[bi]) != before:
                    changed = True
    return follow

# ── Collect terminals ────────────────────────────────────────────────────

def collect_terminals():
    terms = set()
    for _, rhs in PRODS:
        for c in rhs:
            if not is_upper(c) and c != '#':
                terms.add(c)
    terms.add('$')
    return sorted(terms)

# ── Build LL(1) Table ────────────────────────────────────────────────────

def build_table(first, follow):
    terms = collect_terminals()
    # table[nt][terminal] = production index or -1
    table = {nt: {t: -1 for t in terms} for nt in NONTERMS}
    is_ll1 = True

    for i, (lhs, rhs) in enumerate(PRODS):
        fa = first_of_str(rhs, first)
        for sym in fa:
            if sym == '#':
                continue
            if sym in table[lhs]:
                if table[lhs][sym] != -1 and table[lhs][sym] != i:
                    is_ll1 = False
                table[lhs][sym] = i
        if '#' in fa:
            for sym in follow[nt_index(lhs)]:
                if sym in table[lhs]:
                    if table[lhs][sym] != -1 and table[lhs][sym] != i:
                        is_ll1 = False
                    table[lhs][sym] = i

    return table, terms, is_ll1

# ── Print ────────────────────────────────────────────────────────────────

def print_first_follow(first, follow):
    print("\nFIRST Sets:")
    for i, nt in enumerate(NONTERMS):
        print(f"  FIRST({nt}) = {{ {', '.join(sorted(first[i]))} }}")
    print("\nFOLLOW Sets:")
    for i, nt in enumerate(NONTERMS):
        print(f"  FOLLOW({nt}) = {{ {', '.join(sorted(follow[i]))} }}")

def print_table(table, terms):
    print("\nLL(1) Parsing Table\n")
    header = f"{'NT/T':<6}" + "".join(f"{t:<12}" for t in terms)
    print(header)
    for nt in NONTERMS:
        row = f"{nt:<6}"
        for t in terms:
            p = table[nt][t]
            if p == -1:
                cell = '-'
            else:
                lhs, rhs = PRODS[p]
                cell = f"{lhs}->{rhs}"
            row += f"{cell:<12}"
        print(row)

# ── Main ─────────────────────────────────────────────────────────────────

def main():
    print("Given Grammar:")
    print("S -> aBDh")
    print("B -> cC")
    print("C -> bC | #")
    print("D -> EF")
    print("E -> g | #")
    print("F -> f | #")

    first  = compute_first()
    follow = compute_follow(first)
    table, terms, is_ll1 = build_table(first, follow)

    print_first_follow(first, follow)
    print_table(table, terms)

    status = "Grammar is LL(1)." if is_ll1 else "Grammar is NOT LL(1). (Conflict found)"
    print(f"\nGrammar status: {status}")

if __name__ == "__main__":
    main()
